pg_shard.c

```diff
@@ -713,25 +768,123 @@ ClassifyRestrictions(List *queryRestrictList, List **remoteRestrictList,
 
 
 /*
- * RowAndColumnFilterQuery builds a query which contains the filter clauses from
+ * ClassifyHasAggregates decides whether local and distributed queries will include
+ * aggregates or not. remoteQueryHasAggregates and localQueryHasAggregates are output
+ * parameters to receive these boolean values.
+ */
+static void
+ClassifyHasAggregates(Query *query, bool *remoteQueryHasAggregates,
+					  bool *localQueryHasAggregates)
+{
+	*remoteQueryHasAggregates = false;
+	*localQueryHasAggregates = false;
+
+	if (query->groupClause != NIL)
+	{
+		Oid tableId = ExtractFirstDistributedTableId(query);
+		bool safeToPushDownAggregate = SafeToPushDownAggregate(query, tableId);
+
+		/* if we can push down the aggregates update */
+		if (safeToPushDownAggregate)
+		{
+			*remoteQueryHasAggregates = true;
+			*localQueryHasAggregates = false;
+		}
+		else
+		{
+			*remoteQueryHasAggregates = false;
+			*localQueryHasAggregates = true;
+		}
+	}
+}
+
+
+/*
+ * SafeToPushDownAggregate returns true if GROUP BY clause includes the partition column.
+ * This function returns false if the total length of target list is greater than the
+ * total number of columns on the distributed table whose Oid is the second parameter
+ * of this function.
+ */
+static bool
+SafeToPushDownAggregate(Query *query, Oid distributedTableId)
+{
+	bool safeToPushDownAggregates = false;
+	List *targetList = query->targetList;
+	List *groupClauseList = query->groupClause;
+	List *rangeTableList = query->rtable;
+	RangeTblEntry *rangeTableEntry = (RangeTblEntry *) linitial(rangeTableList);
+	int tableColumnCount = list_length(rangeTableEntry->eref->colnames);
+	int targetListCount = list_length(targetList);
+
+	/*
+	 * standard_planner will be called on localQuery at PlanSequentialScan function.
```

Comment:
Given the more descriptive function comment, I'd replace this with a simple _return false if there is no GROUP BY clause or it the target list has more elements than the table has columns_.